channels/smartcard/client/smartcard_pack.c

```diff
@@ -79,38 +178,149 @@ static LONG smartcard_ndr_read(wStream* s, BYTE** data, size_t min)
 	if (!r)
 		return SCARD_E_NO_MEMORY;
 	Stream_Read(s, r, len);
+	smartcard_unpack_read_size_align(NULL, s, len, 4);
 	*data = r;
 	return STATUS_SUCCESS;
 }
 
-static LONG smartcard_ndr_read_a(wStream* s, CHAR** data, size_t min)
+static BOOL smartcard_ndr_pointer_write(wStream* s, UINT32* index, DWORD length)
+{
+	const UINT32 ndrPtr = 0x20000 + (*index) * 4;
+
+	if (!s)
+		return FALSE;
+	if (!Stream_EnsureRemainingCapacity(s, 4))
+		return FALSE;
+
+	if (length > 0)
+	{
+		Stream_Write_UINT32(s, ndrPtr); /* mszGroupsNdrPtr (4 bytes) */
+		(*index) = (*index) + 1;
+	}
+	else
+		Stream_Write_UINT32(s, 0);
+	return TRUE;
+}
+
+static LONG smartcard_ndr_write(wStream* s, const BYTE* data, UINT32 size, UINT32 elementSize,
+                                ndr_ptr_t type)
+{
+	const UINT32 offset = 0;
+	const UINT32 len = size;
+	const UINT32 dataLen = size * elementSize;
+	size_t required;
+
+	if (size == 0)
+		return SCARD_S_SUCCESS;
+
+	switch (type)
+	{
+		case NDR_PTR_FULL:
+			required = 12;
+			break;
+		case NDR_PTR_SIMPLE:
+			required = 4;
+			break;
+		case NDR_PTR_FIXED:
+			required = 0;
+			break;
+	}
+
+	if (!Stream_EnsureRemainingCapacity(s, required + dataLen + 4))
+		return STATUS_BUFFER_TOO_SMALL;
+
+	switch (type)
+	{
+		case NDR_PTR_FULL:
+			Stream_Write_UINT32(s, len);
+			Stream_Write_UINT32(s, offset);
+			Stream_Write_UINT32(s, len);
+			break;
+		case NDR_PTR_SIMPLE:
+			Stream_Write_UINT32(s, len);
+			break;
+		case NDR_PTR_FIXED:
+			break;
+	}
+
+	if (data)
+		Stream_Write(s, data, dataLen);
+	else
+		Stream_Zero(s, dataLen);
+	return smartcard_pack_write_size_align(NULL, s, len, 4);
+}
+
+static LONG smartcard_ndr_write_state(wStream* s, const ReaderState_Return* data, UINT32 size,
+                                      ndr_ptr_t type)
+{
+	union {
+		const ReaderState_Return* reader;
+		const BYTE* data;
+	} cnv;
+
+	cnv.reader = data;
```

Comment:
@akallabeth I don't see in these 3 functions why we need a union, can you elaborate ?

Reply:
this is the only way in C to cast away const ;)